### PR TITLE
Metric Logging

### PR DESCRIPTION
**Pull request by calvinpelletier**
Build metric logger: 
```python
metric_logger = get_metric_logger(**config.metrics)
````

Log a metric:
```python
metric_logger.log("name", value, step)
```

If you want to check if this is a log step before calculating the metric:
```python
if metric_logger.is_log_step(name, step):
    value = calculate_metric()
    metric_logger.log(name, value, step)
```

Open questions:
- multiple loggers? e.g. both wandb and stdout
- non-constant log frequencies? (e.g. for metrics that take a long time to calculate, I like to use a logarithmic frequency so that i get quick data points early in training and fewer points as training becomes more stable)  

Follow up PRs:
- put metric logger inside an actor

### CODE REVIEW

File: apps/sft/llama3_8b.yaml [hunk 1]
```diff
@@ -2,10 +2,10 @@
 # profiling:
 #   enable_profiling: false
 
-# metrics:
-#   log_freq: 10
-#   enable_tensorboard: true
-#   save_tb_folder: "tb"
+metrics:
+  logger: "stdout"
```
Comment (ebsmothers) [inline apps/sft/llama3_8b.yaml]: nit: don't need the quotes here (I think)

File: apps/sft_v2/main.py [hunk 2]
```diff
@@ -74,7 +75,7 @@ class ForgeSFTRecipe(ForgeActor, ForgeEngine):
     def __init__(self, job_config: ForgeJobConfig):
         self.current_step = 0
         self.num_training_steps = job_config.training.steps
-        self.metric_logger = None  # TODO: fix this
+        self.metric_logger = get_metric_logger(**job_config.metrics)
```
Comment (ebsmothers) [inline apps/sft_v2/main.py]: Will this work in the Monarch-style SFT script? (Worth testing cause I thought for that case we wanted to wrap in an Actor)

File: src/forge/util/logging.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    Args:
+        logger (logging.Logger): The logger.
+        msg (str): The warning message.
+        level (int): The logging level. See https://docs.python.org/3/library/logging.html#levels for values.
+            Defaults to ``logging.INFO``.
+    """
+    log_rank_zero(logger=logger, msg=msg, level=level)
+
+
+def deprecated(msg: str = "") -> Callable[[T], T]:
```
Comment (ebsmothers) [inline src/forge/util/logging.py]: Hopefully we don't need anything from here on down yet.. I would just remove it

File: src/forge/types.py [hunk 4]
```diff
@@ -98,3 +101,6 @@ class ProcessConfig:
     oncall: str = "torchtune"
     identity: str = "pytorch_distributed"
     image: str = "forge_workspace:latest"
+
+
+Scalar = Union[Tensor, ndarray, int, float]
```
Comment (ebsmothers) [inline src/forge/types.py]: nit but I wouldn't call it a Scalar (since a Tensor is kinda by definition not a scalar)

File: src/forge/interfaces.py [hunk 5]
```diff
@@ -150,3 +150,110 @@ def tokenize_messages(
...
+    Args:
+        log_freq (Mapping[str, int]):
+            calls to `log` and `log_dict` will be ignored if `step % log_freq[metric_name] != 0`
+    """
+
+    def __init__(self, log_freq: Mapping[str, int]):
+        self._log_freq = log_freq
+        self._step = None
+
+    def set_step(self, step: int) -> None:
```
Comment (ebsmothers) [inline src/forge/interfaces.py]: Is there a specific reason we want to enable calls to log with optional step argument? Imo it makes the logic a bit more complicated, it would be nice to just guarantee this is being passed all the time.

File: src/forge/interfaces.py [hunk 6]
```diff
@@ -150,3 +150,110 @@ def tokenize_messages(
...
+            step (int): current step. if not given, will use the one last provided via set_step()
+        """
+        if step is None:
+            assert (
+                self._step is not None
+            ), "`step` arg required if `set_step` has not been called."
+            step = self._step
+        return step % self._log_freq[name] == 0
+
+    def log(
```
Comment (ebsmothers) [inline src/forge/interfaces.py]: Personally I would lean towards keeping implementations out of the ABC as much as possible. Especially cause in this case (1) we are duplicating `log`/`_log` and `log_dict`/`_log_dict`, all pretty much as syntactic sugar to handle the logic of `step % log_freq == 0`.

File: src/forge/util/metric_logging.py [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+
+# TODO: replace with direct instantiation via a path to the class in the config
+METRIC_LOGGER_STR_TO_CLS = {
+    "stdout": StdoutLogger,
+    "tensorboard": TensorBoardLogger,
+    "wandb": WandBLogger,
+}
+
+
+def _get_rank():
```
Comment (ebsmothers) [inline src/forge/util/metric_logging.py]: We can just use [get_world_size_and_rank](https://github.com/pytorch/torchtune/blob/main/torchtune/training/_distributed.py#L306) and put in util/distributed.py or something

File: src/forge/interfaces.py (hunk 5 above)
Comment (calvinpelletier) [reply src/forge/interfaces.py]: I personally like this style of metric logging because then you can just notify the logger of the step once in the main training loop and then all the rest of your metric code doesn't need to worry about passing around a step variable.

But I can change it to keep the logger as simple as possible

File: apps/sft_v2/main.py (hunk 2 above)
Comment (calvinpelletier) [reply apps/sft_v2/main.py]: I was originally going to do that in a follow-up PR. I wasn't able to get the monarch-style sft script working even without my changes

File: src/forge/types.py (hunk 4 above)
Comment (calvinpelletier) [reply src/forge/types.py]: yeah this came from torchtune code, I assume it was signifying that the tensor/ndarrays would only have one value in them. I'll remove it and enforce int or float

File: src/forge/util/metric_logging.py [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+        organize_logs (bool): If `True`, this class will create a subdirectory within `log_dir` for the current
+            run. Having sub-directories allows you to compare logs across runs. When TensorBoard is
+            passed a logdir at startup, it recursively walks the directory tree rooted at logdir looking for
+            subdirectories that contain tfevents data. Every time it encounters such a subdirectory,
+            it loads it as a new run, and the frontend will organize the data accordingly.
+            Recommended value is `True`. Run `tensorboard --logdir my_log_dir` to view the logs.
+        **kwargs: additional arguments
+
+    Example:
+        >>> from torchtune.training.metric_logging import TensorBoardLogger
```
Comment (ebsmothers) [inline src/forge/util/metric_logging.py]: Make sure to update the docstrings 😃 (I notice something similar in WandBLogger too)

File: apps/sft/llama3_8b.yaml [hunk 9]
```diff
@@ -2,10 +2,11 @@
 # profiling:
 #   enable_profiling: false
 
-# metrics:
-#   log_freq: 10
-#   enable_tensorboard: true
-#   save_tb_folder: "tb"
+metrics:
```
Comment (ebsmothers) [inline apps/sft/llama3_8b.yaml:5]: Probably the proper long-term thing to do is to define our own class [here](https://github.com/pytorch/torchtitan/blob/7d744b2318b279f9ae3afbec5a5f1f53dd740964/torchtitan/experiments/forge/job_config.py#L25) (it'll be helpful to have a single source of truth for our various config fields). But doesn't have to block this PR

File: apps/sft/main.py [hunk 10]
```diff
@@ -60,7 +61,7 @@ def __init__(self, job_config: ForgeJobConfig):
         self.num_training_steps = job_config.training.steps
         self.gradient_accumulation_steps = 1  # Example value, adjust as needed
         super().__init__(job_config)
-        self.metric_logger = None  # TODO: fix this
+        self.metric_logger = get_metric_logger(**job_config.metrics)
```
Comment (ebsmothers) [inline apps/sft/main.py:67]: Related to my other comment -- once we have a proper `job_config.metrics` field we could drop the unpack here and just take `job_config.metrics` directly here. This would be more consistent with how APIs are called in ForgeEngine

File: apps/sft/llama3_8b.yaml [hunk 11]
```diff
@@ -2,10 +2,11 @@
...
+  logger: wandb
+  project: dsafjkdsafjlskdjf
```
Comment (ebsmothers) [inline apps/sft/llama3_8b.yaml]: Let's remove this (is project actually necessary for WandB? I thought they would create one for you if you don't pass it).

File: apps/sft/llama3_8b.yaml [hunk 12]
```diff
@@ -2,10 +2,11 @@
 # profiling:
 #   enable_profiling: false
 
-# metrics:
-#   log_freq: 10
-#   enable_tensorboard: true
-#   save_tb_folder: "tb"
+metrics:
+  logger: wandb
```
Comment (ebsmothers) [inline apps/sft/llama3_8b.yaml]: Do we want this to be the default logger? Personally I prefer it but (a) not sure it works on MAST (may need tensorboard there) and (b) we may want to add it as a dependency

Comment (calvinpelletier) [reply apps/sft/llama3_8b.yaml]: yeah probably best for tensorboard to be the default unless we decide on adding wandb as a dep

**Comment by calvinpelletier** [comment]
@ebsmothers addressed the comments. I'll leave the config class for the next PR after adding metric logging for sft_v2 and toy_rl

File: src/forge/util/logging.py [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+    """
+    Get a logger with a stream handler.
+
+    Args:
+        level (Optional[str]): The logging level. See https://docs.python.org/3/library/logging.html#levels for list of levels.
+
+    Example:
+        >>> logger = get_logger("INFO")
+        >>> logger.info("Hello world!")
+        INFO:torchtune.utils._logging:Hello world!
```
Comment (ebsmothers) [inline src/forge/util/logging.py:26]:
```suggestion
        INFO:forge.util.logging:Hello world!
```

File: src/forge/util/metric_logging.py [hunk 14]
```diff
@@ -0,0 +1,285 @@
...
+
+class WandBLogger(MetricLogger):
+    """Logger for use w/ Weights and Biases application (https://wandb.ai/).
+    For more information about arguments expected by WandB, see https://docs.wandb.ai/ref/python/init.
+
+    Args:
+        freq (Mapping[str, int]):
+            calls to `log` and `log_dict` will be ignored if `step % freq[metric_name] != 0`
+        log_dir (Optional[str]): WandB log directory.
+        project (str): WandB project name. Default is `torchtune`.
```
Comment (ebsmothers) [inline src/forge/util/metric_logging.py:174]:
```suggestion
        project (str): WandB project name.
```

File: src/forge/util/logging.py [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+
+    Args:
+        logger (logging.Logger): The logger.
+        msg (str): The warning message.
+        level (int): The logging level. See https://docs.python.org/3/library/logging.html#levels for values.
+            Defaults to ``logging.INFO``.
+    """
+    rank = dist.get_rank() if dist.is_available() and dist.is_initialized() else 0
+    if rank != 0:
+        return
```
Comment (DNXie) [inline src/forge/util/logging.py:52]: Shouldn't log rank be configurable like how [titan](https://github.com/pytorch/torchtitan/blob/ad06609c4e3ac119112b097b6b317f31c5c760f5/run_train.sh#L14) did? But I guess we can always improve later.
